### PR TITLE
Config updates (size, etc)

### PR DESCRIPTION
**Pull request by joshmoore**
Add support for all the values in the following config file (defaults are shown):

```
$ cat ~/.starfish.config
{
  "backend": {
    "caching": {
      "debug": false,
      "directory": "~/.starfish-cache",
      "size_limit": 5e9
    },
  },
  "validation": {
    "strict": false
  }
}
$ export STARFISH_CONFIG="@~/.starfish.config"
```

follows on: https://github.com/spacetx/starfish/pull/645
depends on: https://github.com/spacetx/slicedimage/pull/76

## Test plan: ##

Add the above config file locally updating `directory` as wished and set `STARFISH_CONFIG`. All subsequent access of HTTP-based filesets should be cached to the new directory.

## Questions ##

 * [x] Should ~/.starfish-config (or just ~/.starfish?) be read by default?
 * [x] Should `["backend"]` actually be `["slicedimage"]`?

### CODE REVIEW

**Comment by joshmoore** [comment]
:+1: Few questions over there, then I'll get both cleaned up.

File: starfish/util/config.py [hunk 1]
```diff
@@ -35,10 +35,13 @@ def __init__(self, value: Union[str, Dict]=None,
         if not value:
             value = {}
 
+        data: Dict = {}
```
Comment (ttung) [inline starfish/util/config.py]:
```suggestion
        data: Dict
```

File: starfish/test/test_config.py [hunk 2]
```diff
@@ -50,3 +52,49 @@ def test_lookup_deep():
...
+        "backend": {
+            "cache": {
+                "enabled": True,
+            }}}),
+    ("disabled", {
+        "expected": 0,
+        "backend": {
+            "cache": {
+                "enabled": False,
+            }}}),
```
Comment (ttung) [inline starfish/test/test_config.py]: would be super cool to have a test case for a cache that's enabled but size-limited.

File: starfish/experiment/experiment.py [hunk 3]
```diff
@@ -189,24 +192,27 @@ def from_json(cls,
             Experiment object serving the requested experiment data
 
         """
+
+        config_obj = Config(config)  # STARFISH_CONFIG is assumed
+        backend_config = config_obj.lookup(["backend"], {})
```
[the remark this answers is not in the document]
Comment (ttung) [reply starfish/experiment/experiment.py]: Also, now that I've seen the update to https://github.com/spacetx/slicedimage/pull/76, I think it makes a lot of sense to do this:

`backend_config = config_obj.lookup(("backend",), {'directory': "~/.starfish-cache"})`

It moves all the starfish-isms into starfish.

[the remark this answers is not in the document]
Comment (joshmoore) [reply starfish/experiment/experiment.py]: _Sweet_

File: starfish/experiment/experiment.py [hunk 4]
```diff
@@ -177,7 +177,10 @@ def from_json(cls,
         STARISH_CONFIG :
             This parameter is read from the environment to permit setting configuration
             values either directly or via a file. Keys read include:
-             - cache.allow_caching
+             - cache.enabled
```
[the remark this answers is not in the document]
Comment (joshmoore) [reply starfish/experiment/experiment.py]: Indeed.

File: starfish/test/test_config.py (hunk 2 above)
Comment (joshmoore) [reply starfish/test/test_config.py]: will do.

File: starfish/experiment/experiment.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshmoore) [reply starfish/experiment/experiment.py]: The downside being that if you set _any_ `backend` config, then you need to set the cache directory.

File: starfish/test/test_config.py (hunk 2 above)
Comment (joshmoore) [reply starfish/test/test_config.py]: test added.

File: starfish/experiment/experiment.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshmoore) [reply starfish/experiment/experiment.py]: Moving to size_limit=0 to disable

**Comment by joshmoore** [comment]
This is now waiting on a release & bump of slicedimage.

**Comment by joshmoore** [comment]
Green 💚 (See 2 questions in the description)

**Comment by ttung** [comment]
> Should ~/.starfish-config (or just ~/.starfish?) be read by default?

Ideally, we should read from `~/.starfish/config`, and the cache should default to `~/.starfish/cache`
Could do in a different PR.

> Should ["backend"] actually be ["slicedimage"]?

That's a good idea too.

**Comment by joshmoore** [comment]
Pushed. (Questions answered in description). The only issue I see with the move of the cache is that there will be a _temporary_ slowdown for travis and early adopters will have a possibly 1GB cache directory that they need to delete (@dganguli, @berl, @ambrosejcarr).

**Comment by ambrosejcarr** [comment]
I'm catching up on work that I missed while planning -- is the crux of this PR adding in an object to support and unify our various configuration modes (environment, (config file?))?

**Comment by joshmoore** [comment]
Yes. After a good deal of back and forth, it looks like the best strategy based on the current codebase is:

 * User-space
   -  If `STARFISH_CONFIG` is set, load that
   - else if `~/.starfish/config` is a valid json file, read it.
   - Also: check individual properties to override (e.g. `STARFISH_VALIDATION_STRICT`)
 * Dev-space
   - Anywhere you need to configure something, do: `StarfishConfig().some_property`
   - i.e. don't try to pass parameters around.

**Pull request merged**